### PR TITLE
[bug545343] Avoid validation of projects without Xtext nature

### PR DESCRIPTION
**Pull request by szarnekow**

### CODE REVIEW

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 1]
```diff
@@ -94,8 +94,12 @@
 	@Inject 
 	private OperationCanceledManager operationCanceledManager;
 	
+	private ClosedProjectsQueue closedProjectsQueue;
+	
 	@Inject
-	private ReflectExtensions reflector;
+	private void injectClosedProjectsQueue(ISharedStateContributionRegistry sharedState) {
+		this.closedProjectsQueue = sharedState.getSingleContributedInstance(ClosedProjectsQueue.class);
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: why is this needed here and not in ProjectOpenedOrClosedListener?

Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: The projectOpenOrClosedListener is instantiated from the same child injector. The build is from the parent injector.

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 2]
```diff
@@ -401,12 +386,17 @@ protected void fullBuild(final IProgressMonitor monitor, boolean isRecoveryBuild
...
 
-		IResourceDeltaVisitor visitor = createDeltaVisitor(toBeBuiltComputer, toBeBuilt, progress);
-		processedAbsentReferencedProjects(visitor);
-		doBuild(toBeBuilt, progress.split(8), 
-			isRecoveryBuild 
-				? BuildType.RECOVERY 
-				: BuildType.FULL);
+		Task task = closedProjectsQueue.exhaust();
+		try {
+			addInfosFromTask(task, toBeBuilt);
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: does this make sense in the recovery usecase too?

Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: The task contains information about the things that do no longer have the XtextBuilder attached. It will ensure that all the remaining entries are indeed removed from the Xtext index. So yes, I think so.

**Comment by cdietrich** [comment]
@szarnekow any idea if the test failures are related?

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java (hunk 2 above)
Comment (cdietrich) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: 👍

**Comment by szarnekow** [comment]
I can't find test failures.

**Comment by cdietrich** [comment]
ahhh i klicked the wrong job sry

**Comment by cdietrich** [comment]
created https://ci.eclipse.org/xtext/job/xtext-xtend/job/bug545343/ for downstream

**Comment by szarnekow** [comment]
@cdietrich What do you think?

**Comment by cdietrich** [comment]
looks per so ok. but i have still no clue on how to reproduce this exactly

**Comment by szarnekow** [comment]
> how to reproduce this exactly

ProjectA - no Xtext nature
ProjectB depends on A - has Xtext nature

Put a dsl file B into src folder of ProjectB (with syntax error) - error markers appears after clean build
Put a dsl file A into src folder of ProjectA (also with syntax error) - no marker may appear on file A

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 3]
```diff
@@ -308,45 +312,26 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
...
+		Task task = closedProjectsQueue.exhaust();
 		try {
-			return (T) reflector.invoke(this, methodName);
-		} catch (SecurityException | IllegalArgumentException | IllegalAccessException | InvocationTargetException
-				| NoSuchMethodException e) {
-			throw new RuntimeException(e);
+			addInfosFromTask(task, toBeBuilt);
+			doBuild(toBeBuilt, progress.split(8), BuildType.INCREMENTAL);
+		} catch(Exception e) {
+			task.reschedule();
```
Comment (cdietrich) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: are we sure this wont produce infinite loops in some circumstances?

Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: I am quite sure... but so was I with the previous broken logic ... i will think about a guard, maybe a reschedule counter

Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: I am quite sure... but so was I with the previous broken logic ... i will think about a guard, maybe a reschedule counter

**Comment by cdietrich** [comment]
we will provide a service release for this issues

**Comment by cdietrich** [comment]
@szarnekow we should think about the counter anyway

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: Done

**Comment by kthoms** [comment]
Please use `[bug#545343]` (add hashtag) to commit message

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 4]
```diff
@@ -401,12 +386,17 @@ protected void fullBuild(final IProgressMonitor monitor, boolean isRecoveryBuild
...
-				? BuildType.RECOVERY 
-				: BuildType.FULL);
+		Task task = closedProjectsQueue.exhaust();
+		try {
+			addInfosFromTask(task, toBeBuilt);
+			doBuild(toBeBuilt, progress.split(8), 
+				isRecoveryBuild 
+					? BuildType.RECOVERY 
+					: BuildType.FULL);
+		} catch(Exception e) {
```
Comment (kthoms) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: nitpick: add whitespace after catch

File: org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java [hunk 5]
```diff
@@ -455,7 +445,14 @@ protected void clean(IProgressMonitor monitor) throws CoreException {
 			if (monitor.isCanceled()) {
 				throw new OperationCanceledException();
 			}
-			doClean(toBeBuilt, progress.split(8));
+			Task task = closedProjectsQueue.exhaust();
+			try {
+				addInfosFromTask(task, toBeBuilt);
+				doClean(toBeBuilt, progress.split(8));
+			} catch(Exception e) {
```
Comment (kthoms) [inline org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java]: nitpick: add whitespace after catch

**Comment by cdietrich** [comment]
restarted jenkins. otherwise ok
